### PR TITLE
infra: deploy to AWS

### PR DESCRIPTION
**Pull request by leite08**
Ref: https://github.com/metriport/metriport-internal/issues/228

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/113
- Downstream:
  - https://github.com/metriport/metriport-internal/pull/371
  - https://github.com/metriport/fhir-server/pull/6

### Description

- Create infra on AWS to publish ECR image
- Create script to build and publish Docker image on AWS ECR
- Create infra on AWS to deploy the FHIR server
- Update README
- CI/CD  done on the downstream PR https://github.com/metriport/fhir-server/pull/6

### Release Plan

- already deployed on `staging`, when the PR is ready to review it can be merged into `develop` at any time
- the merge into `master` needs to be coordinated
  - upstream needs to be merged/deployed into production before this
  - only after https://github.com/metriport/metriport/pull/113 is deployed in `production` we can proceed merging this and downstream dependencies

### CODE REVIEW

File: infra/config/example.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import { EnvType } from "../lib/env-type";
+
+export const config: EnvConfig = {
+  environmentType: EnvType.production,
+  region: "us-east-1",
+  vpcId: "my_vpcId",
+  dbName: "my_db",
+  dbUsername: "my_db_user",
+  fhirServerECRName: "fhir-server-ecr-name",
+};
```
Comment (Goncharo) [inline infra/config/example.ts:11]: assuming there'll be another PR to put the actual config in the internal repo?

File: infra/lib/fhir-server-stack.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    });
+
+    // Retrieve a Docker image from Amazon Elastic Container Registry (ECR)
+    const dockerImage = ecs.ContainerImage.fromEcrRepository(
+      ecr.Repository.fromRepositoryName(
+        this,
+        "FHIRServerECR",
+        props.config.fhirServerECRName
+      )
+    );
```
Comment (Goncharo) [inline infra/lib/fhir-server-stack.ts:139]: why don't we just do what we do in the api stack?

https://github.com/metriport/metriport/blob/66a93687f221abe66ed89e9d018125f904fcfd42/infra/lib/api-stack.ts#L165-L167

this will build & upload to ECR for us

File: infra/lib/fhir-server-stack.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            SPRING_PROFILES_ACTIVE: props.config.environmentType,
+            DB_URL: dbUrl,
+            DB_USERNAME: dbCreds.username,
+          },
+        },
+        healthCheckGracePeriod: Duration.seconds(60),
+        publicLoadBalancer: false,
+        runtimePlatform: {
+          cpuArchitecture: ecs.CpuArchitecture.ARM64,
+          operatingSystemFamily: ecs.OperatingSystemFamily.LINUX,
```
Comment (Goncharo) [inline infra/lib/fhir-server-stack.ts:174]: 🤘

File: src/main/resources/application-local.yaml [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    ### to determine the FHIR server address
+    #   use_apache_address_strategy: false
+    ### forces the use of the https:// protocol for the returned server address.
+    ### alternatively, it may be set using the X-Forwarded-Proto header.
+    #   use_apache_address_strategy_https: false
+    ### enables the server to host content like HTML, css, etc. under the url pattern of /static/**
+    ### the deepest folder level will be used. E.g. - if you put file:/foo/bar/bazz as value then the files are resolved under /static/bazz/**
+    #staticLocation: file:/foo/bar/bazz
+    ### enable to set the Server URL
+    server_address: https://localhost:8080/oauth/fhir
```
Comment (Goncharo) [inline src/main/resources/application-local.yaml:96]: this shouldn't be `8888`?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: infra/config/example.ts (hunk 1 above)
Comment (leite08) [reply infra/config/example.ts:11]: yep! https://github.com/metriport/metriport-internal/pull/371

File: infra/lib/fhir-server-stack.ts (hunk 2 above)
Comment (leite08) [reply infra/lib/fhir-server-stack.ts:139]: Yeah, I started this on OSS then moved the infra here.

The main reason, IIRC, where:
1. CI/CD to fire the deployment comes from updates on this repo
1. Contents to be deployed come from this repo

We ended up pushing a docker image to ECR, so (2) doesn't play a role on this any more.

I'll think about it, maybe we can force a new deploy from this repo's CI/CD using AWS cli (below). But it would be odd to have a new PR on a diff repo just to make a release of this service, while we could have it all in one (following the idea of grouped repos, like we're doing with monorepos).

Also, I'm not sure just having the Fargate/ECS pointing to the `latest` ECR image will work. I just tried to make a new deployment here, and CDK didn't see any diff in task def, so didn't make a new deployment (nor restart of tasks) - I had to manually restart the service (below). Could need to make the task def to point to the current image's tag (`git rev-parse --short HEAD`), then this would go away (and only on edge cases where we need to make a deployment without a new commit we would need to manually restart it).

Thoughts?

For reference, since I just did this - AWS cli to force a new deployment/reload of existing ECS task def:

```shell
$ aws ecs update-service --force-new-deployment --service <SERVICE> --cluster <CLUSTER> --region <REGION>
```

File: deploy-to-ecr.sh [hunk 5]
```diff
@@ -0,0 +1,86 @@
+#!/bin/bash
+############################################################
+# Help                                                     #
+############################################################
+Help()
+{
+   # Display Help
+   echo "Script used to build and deploy the docker image on ECR"
```
[the remark this answers is not in the document]
Comment (leite08) [reply deploy-to-ecr.sh:8]: Would build the docker image and publish it to ECR make sense as part of CDK? Honest question, I never seen this, and AFAIK kinda extrapolates the goal of CDK which is infra provisioning, whereas building docker and publishing the image is more CI/CD like GH Actions.

**Comment by leite08** [comment]
By the way, the FHIR server is working on staging, I just didn't update our Org on CW to point to `api.staging.metriport.com` instead of `fhir.staging.metriport.com` because of the debugging being done to figure out the issue with Document Retrieval.

If it start to take long, we can let them know the environment was updated and now the URL is different.

**Review by Orta21 — APPROVED** [review]

File: infra/lib/fhir-server-stack.ts (hunk 2 above)
Comment (leite08) [reply infra/lib/fhir-server-stack.ts:139]: I missunderstood your message. Looking into it.

Comment (leite08) [reply infra/lib/fhir-server-stack.ts:139]: Yeah, I started off with the idea of having infra on OSS and the code on FHIR server... then I though, "Ill build the image and push to ECR from the FHIR server's repo, and have the actual infra on the OSS".

When I moved the infra from OSS to FHIR server I forgot to question this and reuse the Docker Image construct straight on the CDK code... 🤦 

Adjusting...

File: deploy-to-ecr.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (leite08) [reply deploy-to-ecr.sh:8]: Changed as per https://github.com/metriport/fhir-server/pull/5#discussion_r1116263029

**Review by Goncharo — APPROVED** [review]
🚢 🚢 🚢
